### PR TITLE
feat: new helper function to concat iterables without needing to specify an initial iterable

### PR DESCRIPTION
**Pull request by RebeccaStevens**
This simply allows for a slightly nicer way to write a pipeline that works on multiple initial inputs.

```ts
import {pipe, concat, mergeIterables} from 'iter-ops';

const input1 = [1, 2, 3, 4, 5];
const input2 = [6, 7, 8, 9, 0];

const classicWay = pipe(
    [],
    concat(input1, input2 ),
    // ...
);

const newAltWay = pipe(
    mergeIterables(input1, input2 ),
    // ...
);
```

### CODE REVIEW

File: src/helpers.ts [hunk 1]
```diff
@@ -162,3 +164,20 @@ export function toIterable<T>(i: any): any {
...
+/**
+ * Merge multiple iterators or iterables in to a single iterable.
+ * Merged inputs are iterated over in the order in which they were specified.
+ *
+ * Note that if you concatenate asynchronous iterables with a synchronous pipeline, they will be processed as simple values.
+ *
+ * @see [[concat]]
+ * @category Sync+Async
+ */
+export function mergeIterables<Ts extends [...unknown[]]>(
```
Comment (RebeccaStevens) [inline src/helpers.ts:177]: There's probably a better name for this.

File: src/helpers.ts [hunk 2]
```diff
@@ -162,3 +164,20 @@ export function toIterable<T>(i: any): any {
...
+    ...iters: {
+        [P in keyof Ts]: Any<Ts[P]>;
+    }
+) {
+    return pipe([], concat<never, Ts>(...iters));
```
Comment (RebeccaStevens) [inline src/helpers.ts:182]: This relies on #110 new type info. CI won't pass until this PR is rebase off that one.

**Comment by RebeccaStevens** [comment]
@vitaly-t Thoughts on this

File: src/helpers.ts (hunk 1 above)
Comment (vitaly-t) [reply src/helpers.ts:231]: If you are looking for a shorter syntax, that would be just one function, without `pipe` at all. Other than that, consistency of using the library is important, I do not want to create deviations from that.

Comment (vitaly-t) [reply src/helpers.ts:231]: @RebeccaStevens  You can add locally, if you use it often enough:

```ts
function merge<A>(i1: A[]): Iterable<A>;
function merge<A, B>(i1: A[], i2: B[]): Iterable<A | B>;
function merge<A, B, C>(i1: A[], i2: B[], i3: C[]): Iterable<A | B | C>;

function merge(...values: unknown[]): unknown {
    return pipe([], concat(...values));
}
```

**Comment by vitaly-t** [comment]
> @vitaly-t Thoughts on this

As per my replies above ;)

**Comment by vitaly-t** [comment]
@RebeccaStevens Would it be more appropriate in this shorter format?

**Comment by vitaly-t** [comment]
b.t.w. a shortcut like that does not require use of the pipeline, i.e. we can use the operator directly:

```ts
function merge(...args: any[]) {
    return concat(...args)([]);
}
```

**Comment by vitaly-t** [comment]
@RebeccaStevens Nothing? :)

**Comment by RebeccaStevens** [comment]
Sorry, forgot about this.
I like the change.

The reason I didn't originally just call the function `merge` is because I didn't want it to seem like it was an operator.

**Comment by vitaly-t** [comment]
> The reason I didn't originally just call the function merge is because I didn't want it to seem like it was an operator.

That was clear, though `mergeIterables` wouldn't help there :smile: Perhaps placing it inside `Core` category, and documenting it clearly would help better :wink:

**Comment by RebeccaStevens** [comment]
Yeah, I think it's all good now 👍

**Comment by vitaly-t** [comment]
> I think it's all good now

Are you sure? I've just tried to add an async test, and it didn't work...

image

**Comment by vitaly-t** [comment]
Wait, I may have tried something wrong... Could you, please, confirm it works with async also, and has the right signature for both, and, maybe, add an async test? ;) And, maybe, a type - test, like for operators?

**Comment by RebeccaStevens** [comment]
Alright will do.

**Comment by vitaly-t** [comment]
While you are doing that, I'm gonna try, once again, to rewrite async `concat` (for #156), since these two things go side by side.

**Comment by vitaly-t** [comment]
@RebeccaStevens Ok, I have finished rewriting `concat`, it is in the master. Waiting for your update here now :wink: so we can have an update for two things related to data concatenation.

**Comment by RebeccaStevens** [comment]
This is a little harder than I thought it would be.
Essentially the issue is we don't know whether to perform the sync or async pipeline when `merge` is called.
Ideally, we'd need to setup a complicated system where we analyze the values of the iterables until we get one that isn't sync; at which point we'd need to switch over to the async pipeline.

The much easier solution would just be to have a sync and an async version of merge.

**Comment by vitaly-t** [comment]
The current version of `concat` handles any combination of sync+async, including values and direct iterators. And if a simple shortcut brings so much ambiguity, it is probably not worth doing. I think that adding explicit `mergeSync` and `mergeAsync` would pollute the library with something that adds too little value.

Perhaps it is best that we leave it then? Alas!

**Pull request closed without merging**